liblwgeom/lwgeom_median.c

```diff
@@ -19,156 +19,247 @@
 
-	curr->x = next.x;
-	curr->y = next.y;
-	curr->z = next.z;
+		curr->x = next.x;
+		curr->y = next.y;
+		curr->z = next.z;
+	}
 
 	return delta;
 }
 
-static POINT3D
-init_guess(const POINT3D* points, uint32_t npoints)
+static POINT4D
+init_guess(const POINT4D* points, uint32_t npoints)
 {
-	POINT3D guess = { 0, 0, 0 };
+	assert(npoints > 0);
+	POINT4D guess = { 0, 0, 0, 0 };
 	uint32_t i;
 	for (i = 0; i < npoints; i++)
 	{
-		guess.x += points[i].x / npoints;
-		guess.y += points[i].y / npoints;
-		guess.z += points[i].z / npoints;
+		guess.x += points[i].x * points[i].m;
+		guess.y += points[i].y * points[i].m;
+		guess.z += points[i].z * points[i].m;
+		guess.m += points[i].m;
 	}
-
+	guess.x /= guess.m;
+	guess.y /= guess.m;
+	guess.z /= guess.m;
 	return guess;
 }
 
-static POINT3D*
-lwmpoint_extract_points_3d(const LWMPOINT* g, uint32_t* ngeoms)
+static POINT4D*
+lwmpoint_extract_points_4d(const LWMPOINT* g, POINT4D* points, uint32_t* npoints, int* input_ok, int* input_empty)
 {
+	assert(npoints != NULL);
+	assert(input_ok != NULL);
+	assert(*input_ok == LW_TRUE);
+	assert(g->ngeoms > 0);
 	uint32_t i;
 	uint32_t n = 0;
-	int is_3d = lwgeom_has_z((LWGEOM*) g);
+	int has_z = lwgeom_has_z((LWGEOM*) g);
+	int has_m = lwgeom_has_m((LWGEOM*) g);
 
-	POINT3D* points = lwalloc(g->ngeoms * sizeof(POINT3D));
 	for (i = 0; i < g->ngeoms; i++)
 	{
+		assert(n < g->ngeoms);
+		assert(n <= i);
 		LWGEOM* subg = lwcollection_getsubgeom((LWCOLLECTION*) g, i);
 		if (!lwgeom_is_empty(subg))
 		{
-			getPoint3dz_p(((LWPOINT*) subg)->point, 0, (POINT3DZ*) &points[n++]);
-			if (!is_3d)
-				points[n-1].z = 0.0; /* in case the getPoint functions return NaN in the future for 2d */
+			*input_empty = LW_FALSE;
+			if (!getPoint4d_p(((LWPOINT*) subg)->point, 0, &points[n]))
+			{
+				lwerror("Geometric median: getPoint4d_p reported failure on point (POINT(%g %g %g %g), number %d of %d in input).", points[n].x, points[n].y, points[n].z, points[n].m, i, g->ngeoms);
+				*input_ok = LW_FALSE;
+				break;
+			}
+			if (has_m)
+			{
+				/* This limitation on non-negativity can be lifted
+				 * by replacing Weiszfeld algorithm with different one.
+				 * Possible option described in:
+				 *
+				 * Drezner, Zvi & O. Wesolowsky, George. (1991).
+				 * The Weber Problem On The Plane With Some Negative Weights.
+				 * INFOR. Information Systems and Operational Research.
+				 * 29. 10.1080/03155986.1991.11732158.
+				 */
+				if (points[n].m < 0)
+				{
+					lwerror("Geometric median input contains points with negative weights (POINT(%g %g %g %g), number %d of %d in input). Implementation can't guarantee global minimum convergence.", points[n].x, points[n].y, points[n].z, points[n].m, i, g->ngeoms);
+					*input_ok = LW_FALSE;
+					break;
+				}
+
+				/* points with zero weight are not going to affect calculation, drop them early */
+				if (points[n].m > DBL_EPSILON)
+				{
+					n++;
+				}
+			}
+			else
+			{
+				n++;
+			}
+		}
+	}
+	if (*input_ok)
```

Comment:
Why not do this on the fly? I saw your note about loop unrolling, but does it actually perform significantly better?

Reply:
There's already branching on `has_m` in main loop, so moved it there. There's no branching on has_z there, so I prefer to keep it separate. Original code said that it's "just in case it returns NaN" - how about adding an assert for it to be 0 on input and removing the Z reset completely?

Reply:
Did it via `assert`. It should be 0 now, and when/if someone changes that they'll see.

Reply:
Assert seems OK. The "just in case" is because the code claims that it will become NaN in the future; see `liblwgeom_internal.h:72`.